### PR TITLE
wip: initial e2e tests

### PR DESCRIPTION
**Pull request by Jonathansumner**
Addressing #22 

- [x] Include more robust and comprehensive setup & catching of edge cases
- [x] Extend coverage to all areas of interest
- [x] Introduce GQL to testing suite

### CODE REVIEW

File: tests/testing.py [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+    db = None
+    db_cursor = None
+    gql_client = None
+
+    @classmethod
+    def setUpClass(cls):
+        validator_mnemonic = "nut grocery slice visit barrel peanut tumble patch slim logic install evidence fiction shield rich brown around arrest fresh position animal butter forget cost"
+        validator_seed_bytes = Bip39SeedGenerator(validator_mnemonic).Generate()
+        validator_bip44_def_ctx = Bip44.FromSeed(validator_seed_bytes, Bip44Coins.COSMOS).DeriveDefaultPath()
+        cls.validator_wallet = LocalWallet(PrivateKey(validator_bip44_def_ctx.PrivateKey().Raw().ToBytes()))
```
Comment (bryanchriswhite) [inline tests/testing.py:36]: Maybe we can factor these out to a helper for now. I think this would improve maintainability and readability.

File: tests/testing.py [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+
+        cfg = NetworkConfig(
+            chain_id="testing",
+            url="grpc+http://localhost:9090",
+            fee_minimum_gas_price=1,
+            fee_denomination="atestfet",
+            staking_denomination="atestfet",
+        )
+
+        gov_cfg = grpc.insecure_channel('localhost:9090')
```
Comment (bryanchriswhite) [inline tests/testing.py:54]: nit: maybe we can call this `gov_client` as assuming it's a grpc client.

File: tests/testing.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+        tx = Transaction()
+        tx.add_message(msg)
+
+        tx = utils.prepare_and_broadcast_basic_transaction(self.ledger_client, tx, self.validator_wallet)
+        tx.wait_to_complete()
+        self.assertTrue(tx.response.is_successful())
+        # response = self.gov_module.Proposal(gov_query.QueryProposalRequest(proposal_id=1))
+
+    def testProposalVote(self):
+        self.db_cursor.execute('TRUNCATE table gov_proposal_votes')
```
Comment (bryanchriswhite) [inline tests/testing.py:118]: Let's add a comment that this test depends on the prior one.

File: tests/testing.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+        self.db.commit()
+        self.assertFalse(self.db_cursor.execute('SELECT * from gov_proposal_votes').fetchall())
```
Comment (bryanchriswhite) [inline tests/testing.py:120]: :100:

File: tests/testing.py [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+
+        tx = utils.prepare_and_broadcast_basic_transaction(self.ledger_client, self.vote_tx, self.validator_wallet)
+        tx.wait_to_complete()
+        self.assertTrue(tx.response.is_successful())
+
+        time.sleep(5)
+
+        row = self.db_cursor.execute('SELECT * from gov_proposal_votes').fetchone()
```
Comment (bryanchriswhite) [inline tests/testing.py:128]: nit: Can we pull the SQL query out into a variable, I think this would improve maintainability.

File: tests/testing.py [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+    def setUp(self):
+        self.msg = gov_tx.MsgVote(
+            proposal_id=1,
+            voter=self.validator_address,
+            option=gov_pb2.VoteOption.VOTE_OPTION_YES
+        )
+        self.vote_tx = Transaction()
+        self.vote_tx.add_message(self.msg)
+
+    def testGovernanceProposal(self):
```
Comment (bryanchriswhite) [inline tests/testing.py:93]: This could probably be combined with `setUp`. I think we can still make assertions in there.

File: tests/testing.py [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+             "reverse_aggregated_allowance_approver_cap": "3000000000000000000000000",
+             "lower_swap_limit": "1",
+             "upper_swap_limit": "1000000000000000000000000",
+             "swap_fee": "0",
+             "paused_since_block": 18446744073709551615,
+             "denom": "atestfet",
+             "next_swap_id": 0
+             },
+            self.validator_wallet
+        )
```
Comment (bryanchriswhite) [inline tests/testing.py:270]: I think we've realized that we don't need to deploy this if we're relying on oneshot bridge-related docker services to perform setup.

File: tests/testing.py [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+class TransactionTests(Base):
+    amount = 5000000
+    denom = "atestfet"
+    msg_type = '/cosmos.bank.v1beta1.MsgSend'
+
+    def testNativeTransfer(self):
+
+        self.db_cursor.execute('TRUNCATE table native_transfers')
+        self.db.commit()
+
```
Comment (bryanchriswhite) [inline tests/testing.py:203]: I like how you've been checking that the tables are empty! :smile:

File: tests/testing.py [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+            {"swap": {"destination": self.validator_address}},
+            self.validator_wallet,
+            funds=str(self.amount)+self.denom
+        )
+
+        time.sleep(12)
+
+        row = self.db_cursor.execute('SELECT * from legacy_bridge_swaps').fetchone()
+        self.assertIsNotNone(row)
+        self.assertEqual(row[1], self.validator_address)
```
Comment (bryanchriswhite) [inline tests/testing.py:287]: Let's assert for the other fields as well.

**Comment by bryanchriswhite** [comment]
Closing as these have been separated and included in other PRs.
